src/metabase/driver/common.clj

```diff
@@ -298,3 +299,15 @@
            (apply max-key val)
            key
            class->base-type)))))
+
+(def ^:private days-of-week
+  [:monday :tuesday :wednesday :thursday :friday :saturday :sunday])
+
+(defn start-of-week-offset
+  "Return the offset for start of week to have the week start on `setting/start-of-week` given  `driver`."
+  [driver]
+  (let [db-start-of-week     (.indexOf ^clojure.lang.PersistentVector days-of-week (driver/db-start-of-week driver))
+        target-start-of-week (.indexOf ^clojure.lang.PersistentVector days-of-week (setting/get-keyword :start-of-week))
+        delta                (int (- target-start-of-week db-start-of-week))]
+    (* (Integer/signum delta)
+       (- 7 (Math/abs delta)))))
```

Comment:
I might have missed them, but I didn't see tests for this. It's probably worth adding some tests around `start-of-week-offset` and the adjust functions. I feel like it'd be easy for an update down the line to introduce an off-by-one error.